sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -813,6 +860,39 @@ object NullPropagation extends Rule[LogicalPlan] {
 }
 
 
+/**
+ * Unwrap the input of IsNull/IsNotNull if the input is NullIntolerant
+ * E.g. IsNull(Not(null)) == IsNull(null)
```

Comment:
This looks uncommon and I'm not sure if there is real case that user writes a predicate like `IsNull(Not(boolean_col))` which is the target of this rule.

The rule looks correct, though.